### DISCUSSION TITLE
Ribbon Screen Reader Support?

### DISCUSSION THREAD

**Discussion by MichaelWDenney** [kirill-grouchnikov/radiance#496] [Q&A]
Does the ribbon interface not support screen readers?  With NVDA running, the ribbon, control titles, and control tooltips are not being read.  The ribbons in, for example, Word are read.

I also tried the Windows 11 screen reader.  I find, in general, it doesn't work as well as NVDA, but it doesn't recognize the ribbon interface either.  With focus on a ribbon control, it states, "No item in view."

**Comment by kirill-grouchnikov** [comment]
This would have to be a community contribution. There might indeed be some parity gaps between Radiance components (command buttons, button strips, the ribbon itself) and core Swing components, and if there are easy wins to close those gaps, I'd be happy to review and merge PRs in the area.

**Reply by MichaelWDenney** [reply]
Just a clarification/correction:  The screen reader *is* working with labels, dropdown lists, checkboxes, bullets, spinners, and a control I rolled myself and added to a ribbon.  Screen reader is not working with task titles, band titles, button titles,  tooltips, anchored controls, and the application menu.

**Reply by kirill-grouchnikov** [reply]
* Task title goes as the command button title of the command toggle button, so the fix would be the same as for button titles.
* Band titles are drawn as text in `RadianceRibbonBandUI.paintBandTitle`. There's probably something along the lines of `JLabel.setText` to set accessible name and description there.
* Anchored controls are command buttons, so the same as button titles.
* And speaking of button titles, probably the same - setting accessible name (as title?) and description (as extra text? or rich tooltip?)
* Application menu is all command buttons
* Tooltips - probably the same as the core Swing tooltips, around accesible name / description
*

**Reply by kirill-grouchnikov** [reply]
It might be a bit more complicated than that. `JButton` has an internal `AccessibleJButton` which extends `AccessibleAbstractButton` to "mirror" the structure of its content (text, icon, state, action, ...). Same for `JLabel` and internal `AccessibleJLabel`. 

To set the expectations - this would have to be a community contribution to cover the full set of Radiance components (command buttons, command button strips, command button panels, command menus, circular progress indicators, switches, and the ribbon).

**Reply by MichaelWDenney** [reply]
Does the focus on [#400](https://github.com/kirill-grouchnikov/radiance/issues/400) mean that, if I work on screen reader support it will not be possible to incorporate it for a few months?

**Reply by kirill-grouchnikov** [reply]
I don't have capacity or inclination to do multiple branches during development. Everything happens in the main branch (except https://github.com/kirill-grouchnikov/radiance/tree/modularization which showed that Radiance isn't ready to get modularized). The work on #400 is going to touch pretty much everything in the `theming` module, and then a lot of things in the `component` module. 

Accessibility support is technically not in the same parts of these two modules, as it is not a part of the rendering pipeline. But it would be in the UI delegates, and those are going to be having a lot of changes. To your question, if you want to have accessibility support earlier, it would have to be on your local copy of Radiance (local, or forked on Github), and then rebased on top of whatever shape the codebase is in after #400 is done.

**Reply by kirill-grouchnikov** [reply]
The only work left done for #400 is documentation and whatever bug fixes will be needed in the next few weeks. It should be stable now to add screen reader support in your fork of 8.0-SNAPSHOT. The target release date is May 30th and that is unlikely to change.

**Reply by MichaelWDenney** [reply]
Fantastic.  I need to start working on the screen reader support in earnest.

Does https://github.com/kirill-grouchnikov/radiance/issues/400 have any impact on the existing aspects of the API or should the transition be seamless?

**Reply by kirill-grouchnikov** [reply]
https://github.com/kirill-grouchnikov/radiance/blob/sunshine/docs/migration-8.0.md

**Reply by MichaelWDenney** [reply]
Just fyi, the screen reader support is still intended to be added sometime this year, but my short-term priorities have been shifted so I don't expect it to occur before you target release date.

**Reply by kirill-grouchnikov** [reply]
No rush on my side, I wouldn't want something this big to be rushed in.

**Reply by MichaelWDenney** [reply]
No.  I was just hoping to start on it now as well as try out your new update.  The delay makes sense, if a little disappointing.
